### DISCUSSION TITLE
tutorials/physics/soft_body

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#231] [User-contributed notes]
# tutorials/physics/soft_body

Soft bodies (or soft-body dynamics) simulate movement, changing shape and other physical properties of deformable objects. This can for example be used to simulate cloth or to create more realistic...

https://docs.godotengine.org/en/4.3/tutorials/physics/soft_body.html

**Comment by zazoopdi** [comment]
Hi team,

Not remotely confident enough to raise a pull request, but this:

> To create pinned joints, select the upper vertices in the **SoftBody** node:

Simply doesn't work in 4.3

To replicate, create a new scene, add a SoftBody3D node, add a mesh then select Pinned Points and attempt to select any vertex in the mesh - nothing happens :(

**Reply by zazoopdi** [reply]
By increasing the Size of the Pinned Points Array, I can _sometimes_ get it to add a point, by repeatedly clicking the desired vertex - hitbox seems extremely precise

**Reply by Radivarig** [reply]
For me after clicking vertices and they turn blue the attachments appear only after reselecting the mesh

**Reply by MrGabr** [reply]
I can pretty consistently pin/unpin vertices when I click and then drag off them.

**Reply by 96Bysc** [reply]
Yes, clicking and dragging the vertices seem to be the only way to consistently pin/unpin. Sometimes they don't show as pinned (they don't show blue, but are already pinned), but moving the mouse over the vertice makes them turn blue (if they are pinned but it doesn't show).

**Reply by BryantCrisos** [reply]
Hi everyone, I've made an addon to help with Soft Body Pinning. Heres the link to the addon:
https://github.com/BryantCrisos/softPin

-Hope this helps!

**Reply by Calinou** [reply]
This is a bug:

- https://github.com/godotengine/godot/issues/108149

**Comment by capuadaniel** [comment]
Hello I'm 3d artist I would like to give suggestions here for future versions.

1 - I would like to have options to automatic clamp the vertex group to my character, it could be done by collision? When I prepare the model I let the vertices inner the model in T position already. A line select would help too, click one vertex them Shift+click other and all them by them shortest way are "selected" (what is actually pinned, but the idea is first select them drag to a surface and maybe have a snap tool...)
2 - Speaking about T or A position, what about deactivate animations automatic while soft body is selected for edition? It would avoid weird and hard to find problems.
3 - Support for .glb format, yes obj is simpler, but glb is already linked with character model data, and the extra data could be erased.
4 - Name indicator, in the case of using .glb format a separate object in a glb file with name + '-soft' could identify the mesh, as we have today with '-col' and '-colonly' for collision mesh or '-loop' for animations, its super useful to make best models, ready to send to our programmer friends and they can just import easily.

I hope these ideas are further debated, worked on, and become positive feedback for development.

**Reply by Calinou** [reply]
Feature proposals should be made on the Godot proposals repository: https://github.com/godotengine/godot-proposals

>  3 - Support for .glb format, yes obj is simpler, but glb is already linked with character model data, and the extra data could be erased.

You should be able to use a glTF scene's mesh by using the **Save to File** option after selecting a mesh in the [Advanced Import settings dialog](https://docs.godotengine.org/en/stable/tutorials/assets_pipeline/importing_3d_scenes/import_configuration.html#using-the-advanced-import-settings-dialog). The saved mesh can then be loaded in a SoftBody3D node. (This works for FBX and Collada too.)

**Comment by jmfp** [comment]
Not that it matters but since nobody has mentioned it yet the final image on this documentation page leaves quite a bit to be desired. It's nearly impossible to distinguish the "cape" from a regular plane

**Reply by Calinou** [reply]
It depends on player movement, but when the player is standing still, the cape will eventually become close to a plane (since most of its vertices are aligned).

If you want it to be less flat in a resting position, you could use a mesh that's pre-distorted, but Godot's primitive meshes can't do that out of the box.

**Reply by jmfp** [reply]
Oh I totally understand that, I was just suggesting maybe use an image of this with the player moving instead of standing still, for exactly the reason you just stated. If this is supposed to be documentation on soft bodys and how you can make things like capes with them, using an image of the "cape" while it instead looks like a plane to convey that message seems counter intuitive.

**Comment by C-Corpze** [comment]
Silly question perhaps but, is it possible to use a lower resolution mesh with softbody physics to deform a high-res mesh?
For optimization reasons you probably do not want to use a mesh that has thousands of polygons for soft bodies but you might still want to deform it.
Maybe you got a character with a fluffy tail that you want to sway around and deform when it collides with stuff, but that character might have 50k+ polygons.

**Reply by Calinou** [reply]
This isn't currently possible out of the box, but maybe with a custom shader.

That said, for character tails, it may be better to use jiggle bone functionality as found in SpringBoneSimulator3D. These generally don't need full-blown soft body simulation.
